components/net/filemanager_thread.rs

```diff
@@ -108,109 +109,205 @@ enum FileImpl {
                     if let Ok(id) = Uuid::parse_str(&id.0) {
-                        self.inc_ref(id, origin);
+                        spawn_named("inc ref".to_owned(), move || {
+                            let _ = store.inc_ref(&id, &origin);
+                        })
                     }
                 }
                 FileManagerThreadMsg::Exit => break,
             };
         }
     }
 
-    fn inc_ref(&mut self, id: Uuid, origin_in: FileOrigin) {
-        match self.store.get(&id) {
+    fn process_request(&self, load_data: LoadData, consumer: LoadConsumer,
+                       rel_pos: RelativePos, id: Uuid) {
+        let origin_in = load_data.url.origin().unicode_serialization();
+        match self.store.get_impl(&id, &origin_in) {
+            Ok(file_impl) => {
+                match file_impl {
+                    FileImpl::Memory(ref buffered) => {
+                        let buffered = buffered.clone();
+                        let classifier = self.classifier.clone();
+                        spawn_named("load blob".to_owned(), move ||
+                            load_blob(load_data, consumer, classifier,
+                                      None, rel_pos, buffered));
+                    }
+                    FileImpl::PathOnly(ref filepath) => {
+                        let opt_filename = filepath.file_name()
+                                               .and_then(|osstr| osstr.to_str())
+                                               .map(|s| s.to_string());
+
+                        let mut bytes = vec![];
+                        let mut handler = File::open(filepath).unwrap();
+                        let mime = guess_mime_type_opt(filepath);
+                        let size = handler.read_to_end(&mut bytes).unwrap();
+
+                        let entry = BlobURLStoreEntry {
+                            type_string: match mime {
+                                Some(x) => format!("{}", x),
+                                None    => "".to_string(),
+                            },
+                            size: size as u64,
+                            bytes: bytes,
+                        };
+                        let classifier = self.classifier.clone();
+                        spawn_named("load blob".to_owned(), move ||
+                            load_blob(load_data, consumer, classifier,
+                                      opt_filename, rel_pos, entry));
+                    },
+                    FileImpl::Sliced(ref id, ref rel_pos) => {
+                        self.process_request(load_data, consumer, rel_pos.clone(), id.clone());
+                    }
+                }
+            }
+            Err(e) => {
+                send_error(load_data.url.clone(), NetworkError::Internal(format!("{:?}", e)), consumer);
+            }
+        }
+    }
+}
+
+struct FileManagerStore<UI: 'static + UIProvider> {
+    entries: RwLock<HashMap<Uuid, FileStoreEntry>>,
```

Comment:
@Manishearth I moved the `RwLock` under management the `store` itself, and control the block on locking locally in the following first few methods.